# PART II   < Python Data Analysis >

## 11. [ Pandas Programming Workshop ] - Python 網路爬蟲 - 台灣證券交易所 (TWSE) 股票資料

_C. Alex Hu, PhD @  2023/01/08_


> **[ REFERENCE ] :**
>
> + **臺灣證券交易所 (Taiwan Stock Exchange, TWSE)** https://www.twse.com.tw/zh/
>
>
> + Vincent Tsai, "**`TWSE_Python_crawler`**" 
>      + [Github] : "**`TWSE_Python_crawler`**" https://github.com/tsaisunghao/TWSE_Python_crawler
>      + [Youtube] : "Python網路爬蟲 爬取台灣證券交易所歷史交易紀錄(加強版)" https://youtu.be/dvw749iveJA

## <  CONTENTS  >

- [1. Crawler (網路爬蟲) - TWSE 個股資料範例](#Crawler)
- [2. 輸出個股資料至 csv 檔案](#tocsv)
- [3. 從 csv 檔案輸入個股資料](#fromcsv)

<a id='Crawler'></a>
## 1. Crawler (網路爬蟲) - TWSE 個股資料範例

 + **臺灣證券交易所 (Taiwan Stock Exchange, TWSE)** https://www.twse.com.tw/zh/
 
 
 + **2330 台積電 (TSMC) 2023/01/19**
     + **http://www.twse.com.tw/exchangeReport/STOCK_DAY?response=json&date=20230119&stockNo=2330**

In [1]:
import bs4 as bs  #  beautifulsoup 4
import urllib
import urllib.request

### 1.1 連線至台灣證交所 (TWSE)，擷取 台積電 (2330) 2023/01 個股資料

In [2]:
url_history = 'http://www.twse.com.tw/exchangeReport/STOCK_DAY?response=json&date=20230119&stockNo=2330'
webpage_history = urllib.request.urlopen(url_history)
web_html = bs.BeautifulSoup(webpage_history, 'html.parser')

In [5]:
base_url = "https://www.twse.com.tw/exchangeReport/STOCK_DAY?response=json&date={:d}{:02d}01&stockNo={:s}"
year = 2024 #input("Please enter year")
month = 2
sid = "2330"
target_url = base_url.format(year,month,sid)
print(target_url)

https://www.twse.com.tw/exchangeReport/STOCK_DAY?response=json&date=2024201&stockNo=2330


In [3]:
print(web_html)

{"stat":"OK","date":"20230119","title":"112年01月 2330 台積電           各日成交資訊","fields":["日期","成交股數","成交金額","開盤價","最高價","最低價","收盤價","漲跌價差","成交筆數"],"data":[["112/01/03","15,311,364","6,871,973,708","446.00","453.50","443.00","453.00","+4.50","22,581"],["112/01/04","20,626,874","9,310,050,329","449.50","455.00","448.50","449.50","-3.50","18,233"],["112/01/05","23,972,099","10,972,616,269","459.00","459.50","455.00","458.50","+9.00","20,752"],["112/01/06","21,313,593","9,745,142,549","455.00","459.50","455.00","458.50"," 0.00","16,635"],["112/01/09","49,186,355","23,352,375,299","468.00","481.00","467.50","481.00","+22.50","57,305"],["112/01/10","34,785,370","16,867,391,241","486.00","487.00","483.00","486.00","+5.00","36,975"],["112/01/11","21,749,124","10,531,314,878","487.00","488.00","482.00","484.50","-1.50","25,457"],["112/01/12","22,416,984","10,901,640,739","487.50","488.00","484.00","486.50","+2.00","25,780"],["112/01/13","81,848,755","41,113,175,651","507.00","509.00","499.00","500.

In [4]:
import json
stock = json.loads(web_html.text)  #  讀取 JSON 資料
# stock

### 1.2 查看 下載資料 : 台積電 (代碼：2330) 2023/01 個股資料

In [5]:
type(stock)  #  查看 下載資料格式： dictionary

dict

In [6]:
stock.keys()  #  查看 dictionary 的 keys

dict_keys(['stat', 'date', 'title', 'fields', 'data', 'notes'])

In [7]:
stock.values()  #  查看 dictionary 的 values

dict_values(['OK', '20230119', '112年01月 2330 台積電           各日成交資訊', ['日期', '成交股數', '成交金額', '開盤價', '最高價', '最低價', '收盤價', '漲跌價差', '成交筆數'], [['112/01/03', '15,311,364', '6,871,973,708', '446.00', '453.50', '443.00', '453.00', '+4.50', '22,581'], ['112/01/04', '20,626,874', '9,310,050,329', '449.50', '455.00', '448.50', '449.50', '-3.50', '18,233'], ['112/01/05', '23,972,099', '10,972,616,269', '459.00', '459.50', '455.00', '458.50', '+9.00', '20,752'], ['112/01/06', '21,313,593', '9,745,142,549', '455.00', '459.50', '455.00', '458.50', ' 0.00', '16,635'], ['112/01/09', '49,186,355', '23,352,375,299', '468.00', '481.00', '467.50', '481.00', '+22.50', '57,305'], ['112/01/10', '34,785,370', '16,867,391,241', '486.00', '487.00', '483.00', '486.00', '+5.00', '36,975'], ['112/01/11', '21,749,124', '10,531,314,878', '487.00', '488.00', '482.00', '484.50', '-1.50', '25,457'], ['112/01/12', '22,416,984', '10,901,640,739', '487.50', '488.00', '484.00', '486.50', '+2.00', '25,780'], ['112/01/13', '81

In [8]:
stock_info = list(stock.values())  #  轉換成 list 格式
stock_info

['OK',
 '20230119',
 '112年01月 2330 台積電           各日成交資訊',
 ['日期', '成交股數', '成交金額', '開盤價', '最高價', '最低價', '收盤價', '漲跌價差', '成交筆數'],
 [['112/01/03',
   '15,311,364',
   '6,871,973,708',
   '446.00',
   '453.50',
   '443.00',
   '453.00',
   '+4.50',
   '22,581'],
  ['112/01/04',
   '20,626,874',
   '9,310,050,329',
   '449.50',
   '455.00',
   '448.50',
   '449.50',
   '-3.50',
   '18,233'],
  ['112/01/05',
   '23,972,099',
   '10,972,616,269',
   '459.00',
   '459.50',
   '455.00',
   '458.50',
   '+9.00',
   '20,752'],
  ['112/01/06',
   '21,313,593',
   '9,745,142,549',
   '455.00',
   '459.50',
   '455.00',
   '458.50',
   ' 0.00',
   '16,635'],
  ['112/01/09',
   '49,186,355',
   '23,352,375,299',
   '468.00',
   '481.00',
   '467.50',
   '481.00',
   '+22.50',
   '57,305'],
  ['112/01/10',
   '34,785,370',
   '16,867,391,241',
   '486.00',
   '487.00',
   '483.00',
   '486.00',
   '+5.00',
   '36,975'],
  ['112/01/11',
   '21,749,124',
   '10,531,314,878',
   '487.00',
   '488.00',
   

In [9]:
stock_info[2]

'112年01月 2330 台積電           各日成交資訊'

In [10]:
stock_info[3]

['日期', '成交股數', '成交金額', '開盤價', '最高價', '最低價', '收盤價', '漲跌價差', '成交筆數']

### 1.3 將 下載資料 轉換成 data frame 資料格式

In [11]:
import pandas as pd
import numpy as np

stock_price = pd.DataFrame(stock_info[4])
stock_price

0           1               2       3       4       5       6  \
0   112/01/03  15,311,364   6,871,973,708  446.00  453.50  443.00  453.00   
1   112/01/04  20,626,874   9,310,050,329  449.50  455.00  448.50  449.50   
2   112/01/05  23,972,099  10,972,616,269  459.00  459.50  455.00  458.50   
3   112/01/06  21,313,593   9,745,142,549  455.00  459.50  455.00  458.50   
4   112/01/09  49,186,355  23,352,375,299  468.00  481.00  467.50  481.00   
5   112/01/10  34,785,370  16,867,391,241  486.00  487.00  483.00  486.00   
6   112/01/11  21,749,124  10,531,314,878  487.00  488.00  482.00  484.50   
7   112/01/12  22,416,984  10,901,640,739  487.50  488.00  484.00  486.50   
8   112/01/13  81,848,755  41,113,175,651  507.00  509.00  499.00  500.00   
9   112/01/16  38,781,925  19,573,968,332  506.00  508.00  503.00  505.00   
10  112/01/17  29,443,522  14,784,661,794  504.00  504.00  499.50  503.00   

         7       8  
0    +4.50  22,581  
1    -3.50  18,233  
2    +9.00  20,752  
3     0.00  16,635  
4   +22.50  57,305  
5    +5.00  36,975  
6    -1.50  25,457  
7    +2.00  25,780  
8   +13.50  93,121  
9    +5.00  48,051  
10   -2.00  27,930

> ### 將 data frame 資料加上 column name

In [12]:
stock_price.columns = stock_info[3]
stock_price

日期        成交股數            成交金額     開盤價     最高價     最低價     收盤價  \
0   112/01/03  15,311,364   6,871,973,708  446.00  453.50  443.00  453.00   
1   112/01/04  20,626,874   9,310,050,329  449.50  455.00  448.50  449.50   
2   112/01/05  23,972,099  10,972,616,269  459.00  459.50  455.00  458.50   
3   112/01/06  21,313,593   9,745,142,549  455.00  459.50  455.00  458.50   
4   112/01/09  49,186,355  23,352,375,299  468.00  481.00  467.50  481.00   
5   112/01/10  34,785,370  16,867,391,241  486.00  487.00  483.00  486.00   
6   112/01/11  21,749,124  10,531,314,878  487.00  488.00  482.00  484.50   
7   112/01/12  22,416,984  10,901,640,739  487.50  488.00  484.00  486.50   
8   112/01/13  81,848,755  41,113,175,651  507.00  509.00  499.00  500.00   
9   112/01/16  38,781,925  19,573,968,332  506.00  508.00  503.00  505.00   
10  112/01/17  29,443,522  14,784,661,794  504.00  504.00  499.50  503.00   

      漲跌價差    成交筆數  
0    +4.50  22,581  
1    -3.50  18,233  
2    +9.00  20,752  
3     0.00  16,635  
4   +22.50  57,305  
5    +5.00  36,975  
6    -1.50  25,457  
7    +2.00  25,780  
8   +13.50  93,121  
9    +5.00  48,051  
10   -2.00  27,930

> ### 將第一個 column name “`日期`” 改成 “`112年01月 2330 台積電 各日成交資訊`” (`title`)

In [13]:
stock_info[3][0] = stock_info[2]
stock_price.columns = stock_info[3]
stock_price

112年01月 2330 台積電           各日成交資訊        成交股數            成交金額     開盤價  \
0                          112/01/03  15,311,364   6,871,973,708  446.00   
1                          112/01/04  20,626,874   9,310,050,329  449.50   
2                          112/01/05  23,972,099  10,972,616,269  459.00   
3                          112/01/06  21,313,593   9,745,142,549  455.00   
4                          112/01/09  49,186,355  23,352,375,299  468.00   
5                          112/01/10  34,785,370  16,867,391,241  486.00   
6                          112/01/11  21,749,124  10,531,314,878  487.00   
7                          112/01/12  22,416,984  10,901,640,739  487.50   
8                          112/01/13  81,848,755  41,113,175,651  507.00   
9                          112/01/16  38,781,925  19,573,968,332  506.00   
10                         112/01/17  29,443,522  14,784,661,794  504.00   

       最高價     最低價     收盤價    漲跌價差    成交筆數  
0   453.50  443.00  453.00   +4.50  22,581  
1   455.00  448.50  449.50   -3.50  18,233  
2   459.50  455.00  458.50   +9.00  20,752  
3   459.50  455.00  458.50    0.00  16,635  
4   481.00  467.50  481.00  +22.50  57,305  
5   487.00  483.00  486.00   +5.00  36,975  
6   488.00  482.00  484.50   -1.50  25,457  
7   488.00  484.00  486.50   +2.00  25,780  
8   509.00  499.00  500.00  +13.50  93,121  
9   508.00  503.00  505.00   +5.00  48,051  
10  504.00  499.50  503.00   -2.00  27,930

> ### 將 index 欄位 改成 “`112年01月 2330 台積電 各日成交資訊`” (`title`)

In [14]:
df_price = stock_price.set_index(stock_price.columns[0])
df_price

成交股數            成交金額     開盤價     最高價  \
112年01月 2330 台積電           各日成交資訊                                               
112/01/03                          15,311,364   6,871,973,708  446.00  453.50   
112/01/04                          20,626,874   9,310,050,329  449.50  455.00   
112/01/05                          23,972,099  10,972,616,269  459.00  459.50   
112/01/06                          21,313,593   9,745,142,549  455.00  459.50   
112/01/09                          49,186,355  23,352,375,299  468.00  481.00   
112/01/10                          34,785,370  16,867,391,241  486.00  487.00   
112/01/11                          21,749,124  10,531,314,878  487.00  488.00   
112/01/12                          22,416,984  10,901,640,739  487.50  488.00   
112/01/13                          81,848,755  41,113,175,651  507.00  509.00   
112/01/16                          38,781,925  19,573,968,332  506.00  508.00   
112/01/17                          29,443,522  14,784,661,794  504.00  504.00   

                                      最低價     收盤價    漲跌價差    成交筆數  
112年01月 2330 台積電           各日成交資訊                                  
112/01/03                          443.00  453.00   +4.50  22,581  
112/01/04                          448.50  449.50   -3.50  18,233  
112/01/05                          455.00  458.50   +9.00  20,752  
112/01/06                          455.00  458.50    0.00  16,635  
112/01/09                          467.50  481.00  +22.50  57,305  
112/01/10                          483.00  486.00   +5.00  36,975  
112/01/11                          482.00  484.50   -1.50  25,457  
112/01/12                          484.00  486.50   +2.00  25,780  
112/01/13                          499.00  500.00  +13.50  93,121  
112/01/16                          503.00  505.00   +5.00  48,051  
112/01/17                          499.50  503.00   -2.00  27,930

> ### 抽取 data frame 資料的特定欄位

In [15]:
df_price['收盤價']

112年01月 2330 台積電           各日成交資訊
112/01/03    453.00
112/01/04    449.50
112/01/05    458.50
112/01/06    458.50
112/01/09    481.00
112/01/10    486.00
112/01/11    484.50
112/01/12    486.50
112/01/13    500.00
112/01/16    505.00
112/01/17    503.00
Name: 收盤價, dtype: object

In [16]:
df_price[['收盤價', '成交金額', '漲跌價差']]

收盤價            成交金額    漲跌價差
112年01月 2330 台積電           各日成交資訊                                
112/01/03                          453.00   6,871,973,708   +4.50
112/01/04                          449.50   9,310,050,329   -3.50
112/01/05                          458.50  10,972,616,269   +9.00
112/01/06                          458.50   9,745,142,549    0.00
112/01/09                          481.00  23,352,375,299  +22.50
112/01/10                          486.00  16,867,391,241   +5.00
112/01/11                          484.50  10,531,314,878   -1.50
112/01/12                          486.50  10,901,640,739   +2.00
112/01/13                          500.00  41,113,175,651  +13.50
112/01/16                          505.00  19,573,968,332   +5.00
112/01/17                          503.00  14,784,661,794   -2.00

##  [ Exercise 1 ]: 
> ### 請下載 中鋼、日月光、鴻海、廣達、富邦 當月份個股資料，並將其轉換成 data frame 資料格式。

<a id='tocsv'></a>
## 2. 輸出個股資料至 csv 檔案

In [17]:
import os.path
mydir ='./'       #  目前的工作目錄
csv_file = os.path.join(mydir, "stock2330_12345678.csv")   #  建立檔案路徑 + 檔案名稱 stock.csv
df_price.to_csv(csv_file, encoding='utf_8_sig')   #  輸出 csv 檔案至 工作目錄 

<a id='fromcsv'></a>
## 3. 從 csv 檔案輸入個股資料

In [18]:
df_price_csv = pd.read_csv(csv_file)
df_price_csv

112年01月 2330 台積電           各日成交資訊        成交股數            成交金額    開盤價  \
0                          112/01/03  15,311,364   6,871,973,708  446.0   
1                          112/01/04  20,626,874   9,310,050,329  449.5   
2                          112/01/05  23,972,099  10,972,616,269  459.0   
3                          112/01/06  21,313,593   9,745,142,549  455.0   
4                          112/01/09  49,186,355  23,352,375,299  468.0   
5                          112/01/10  34,785,370  16,867,391,241  486.0   
6                          112/01/11  21,749,124  10,531,314,878  487.0   
7                          112/01/12  22,416,984  10,901,640,739  487.5   
8                          112/01/13  81,848,755  41,113,175,651  507.0   
9                          112/01/16  38,781,925  19,573,968,332  506.0   
10                         112/01/17  29,443,522  14,784,661,794  504.0   

      最高價    最低價    收盤價  漲跌價差    成交筆數  
0   453.5  443.0  453.0   4.5  22,581  
1   455.0  448.5  449.5  -3.5  18,233  
2   459.5  455.0  458.5   9.0  20,752  
3   459.5  455.0  458.5   0.0  16,635  
4   481.0  467.5  481.0  22.5  57,305  
5   487.0  483.0  486.0   5.0  36,975  
6   488.0  482.0  484.5  -1.5  25,457  
7   488.0  484.0  486.5   2.0  25,780  
8   509.0  499.0  500.0  13.5  93,121  
9   508.0  503.0  505.0   5.0  48,051  
10  504.0  499.5  503.0  -2.0  27,930

> ### 將 index 欄位 改成 第一個欄位的日期資料

In [19]:
df_price_csv = df_price_csv.set_index(df_price_csv.columns[0])
df_price_csv

成交股數            成交金額    開盤價    最高價  \
112年01月 2330 台積電           各日成交資訊                                             
112/01/03                          15,311,364   6,871,973,708  446.0  453.5   
112/01/04                          20,626,874   9,310,050,329  449.5  455.0   
112/01/05                          23,972,099  10,972,616,269  459.0  459.5   
112/01/06                          21,313,593   9,745,142,549  455.0  459.5   
112/01/09                          49,186,355  23,352,375,299  468.0  481.0   
112/01/10                          34,785,370  16,867,391,241  486.0  487.0   
112/01/11                          21,749,124  10,531,314,878  487.0  488.0   
112/01/12                          22,416,984  10,901,640,739  487.5  488.0   
112/01/13                          81,848,755  41,113,175,651  507.0  509.0   
112/01/16                          38,781,925  19,573,968,332  506.0  508.0   
112/01/17                          29,443,522  14,784,661,794  504.0  504.0   

                                     最低價    收盤價  漲跌價差    成交筆數  
112年01月 2330 台積電           各日成交資訊                              
112/01/03                          443.0  453.0   4.5  22,581  
112/01/04                          448.5  449.5  -3.5  18,233  
112/01/05                          455.0  458.5   9.0  20,752  
112/01/06                          455.0  458.5   0.0  16,635  
112/01/09                          467.5  481.0  22.5  57,305  
112/01/10                          483.0  486.0   5.0  36,975  
112/01/11                          482.0  484.5  -1.5  25,457  
112/01/12                          484.0  486.5   2.0  25,780  
112/01/13                          499.0  500.0  13.5  93,121  
112/01/16                          503.0  505.0   5.0  48,051  
112/01/17                          499.5  503.0  -2.0  27,930

##  [ Exercise 2 ]: 
> ### 請將 [ Exercise 1 ] 中，所下載的 中鋼、日月光、鴻海、廣達、富邦 當月份 data frame 資料，輸出至 csv 檔案。